### ISSUE TITLE
🪙 Dealing with API tokens

### ISSUE THREAD

**Issue by denialhaag** [Munich-Quantum-Software-Stack/QDMI#109]
There is currently no defined way to deal with API tokens that may be needed to interact with the underlying API of a device.

Our (planqc's) device needs an API token for job submission. For now, we have hard-coded the API token in our shared library, but this means that our shared library breaks when the token expires.

If a device needs a token only for job submission, it can, in principle, be set with `QDMI_control_set_parameter`. However, devices may also want to use a token for the query interface.

**Comment by ystade** [comment]
@denialhaag Thanks for bringing this up.

I wonder what the best solution is here. At the moment, I see three different options that I want to summarize in the following:
1. Using [`QDMI_control_set_parameter`](https://munich-quantum-software-stack.github.io/QDMI/client_2control_8h.html#a6e3dfe222ae9404308f6a70283daf496) as suggested by @denialhaag already: For this option, one would need to add another [`QDMI_Job_Parameter`](https://munich-quantum-software-stack.github.io/QDMI/common_2types_8h.html#a4d37195d9d0765b10a6ad681304a4512), e.g., `QDMI_JOB_PARAMETER_TOKEN`. Then, the token can be passed to the job as a parameter.
1. Using [`QDMI_session_set_parameter`](https://munich-quantum-software-stack.github.io/QDMI/session_8h.html#a39d34b2a95cd9e2b8d9d6ab551bff078): For this option, one would need to add another [`QDMI_Session_Parameter`](https://munich-quantum-software-stack.github.io/QDMI/common_2types_8h.html#aae2a513b3364705ca75f709215d912f5), e.g., `QDMI_SESSION_PARAMTER_[VENDOR]_TOKEN` where `[VENDOR]` is a placeholder for the vendor of the device as I think that most likely every vendor will have a different token. Even several devices of the same vendor may be accessed with different tokens.
1. Renaming [`QDMI_control_set_parameter`](https://munich-quantum-software-stack.github.io/QDMI/client_2control_8h.html#a6e3dfe222ae9404308f6a70283daf496) to `QDMI_control_set_job_parameter` (which we should perhaps do anyways) and creating a new function `QDMI_control_set_device_parameter` in a similar style as the other setter for parameters. This way, one could set one token for one device. However, for me, it feels slightly awkward to set a parameter of the device, as in my view, the device has properties that I can query but no parameters I can set, but that might also just be a me-problem.

Any opinion is very welcome.

**Comment by AdVetter** [comment]
It might be good to be able to provide arbitrary config parameters for devices. In our example, we now hardcode also the URL to the machine. If this changes we need now to recompile the file and can not just change URL or replace it with a local IP for testing purposes.

**Comment by burgholzer** [comment]
Hey all 👋🏼 

Thanks for creating this issue to discuss this important matter. 

TL;DR at the bottom 📝 

Let me try to collect my thoughts in the following:

> @denialhaag For now, we have hard-coded the API token in our shared library, but this means that our shared library breaks when the token expires.

I'd generally not advise to statically compile in the API token for various reasons:
- Anyone with access to the shared library could now submit jobs to the device.
- If the token expires, the shared library needs to be recompiled.
- There is no way to have multiple users accessing the device with their token if the token is embedded in the shared library

Hence, I strongly believe an alternative solution is needed.

> @ystade Using [QDMI_control_set_parameter](https://munich-quantum-software-stack.github.io/QDMI/client_2control_8h.html#a6e3dfe222ae9404308f6a70283daf496) as suggested by @denialhaag already: For this option, one would need to add another [QDMI_Job_Parameter](https://munich-quantum-software-stack.github.io/QDMI/common_2types_8h.html#a4d37195d9d0765b10a6ad681304a4512), e.g., QDMI_JOB_PARAMETER_TOKEN. Then, the token can be passed to the job as a parameter.

I think this is something we can do without much effort. It is literally adding a single Enum value. However, as noted by @denialhaag, this probably only gets us so far in the sense that it would only cover the functions involving a `QDMI_Job`.
I'd like to avoid having multiple ways of setting a particular property in QDMI.
Given the points below, I'd argue for not implementing this as a job parameter.

> @ystade Using [QDMI_session_set_parameter](https://munich-quantum-software-stack.github.io/QDMI/session_8h.html#a39d34b2a95cd9e2b8d9d6ab551bff078): For this option, one would need to add another [QDMI_Session_Parameter](https://munich-quantum-software-stack.github.io/QDMI/common_2types_8h.html#aae2a513b3364705ca75f709215d912f5), e.g., QDMI_SESSION_PARAMTER_[VENDOR]_TOKEN where [VENDOR] is a placeholder for the vendor of the device as I think that most likely every vendor will have a different token. Even several devices of the same vendor may be accessed with different tokens.

I would likely prefer to steer clear of this approach, as I believe it would not be effective. Not only would this create quite a few hard-coded enum values, it would also complicate the method for getting the devices from a session as that method would have to match the token parameters to the shared libraries the driver has access to. Furthermore, adding a token after receiving a device would be impossible.

> @ystade Renaming [QDMI_control_set_parameter](https://munich-quantum-software-stack.github.io/QDMI/client_2control_8h.html#a6e3dfe222ae9404308f6a70283daf496) to QDMI_control_set_job_parameter 

I think we should do this anyway (in a separate PR) as I'd consider this an oversight from us in following the typical naming convention.

> @ystade creating a new function QDMI_control_set_device_parameter in a similar style as the other setter for parameters. This way, one could set one token for one device. However, for me, it feels slightly awkward to set a parameter of the device, as in my view, the device has properties that I can query but no parameters I can set, but that might also just be a me-problem.

> @AdVetter It might be good to be able to provide arbitrary config parameters for devices. In our example, we now hardcode also the URL to the machine. If this changes we need now to recompile the file and can not just change URL or replace it with a local IP for testing purposes.

I believe the use case Adrian is mentioning here, along with the ability to set tokens for the device, are already compelling enough to justify the addition of such a feature.
I could see opt-in functionality for certain devices that one would have to enable as another use case.
Also, I believe that recompilation of the device library should not be required for changing things like the base URL or a token.
Side note: such a new function should also be available if a device is opened in read-only mode, as it might be required for a functioning device.

---

I'd propose to

- [ ] Rename [QDMI_control_set_parameter](https://munich-quantum-software-stack.github.io/QDMI/client_2control_8h.html#a6e3dfe222ae9404308f6a70283daf496) to `QDMI_control_set_job_parameter`
- [ ] Introduce a new `QDMI_DEVICE_PARAMETER_T` enum with two pre-defined values `QDMI_DEVICE_PARAMETER_TOKEN` and `QDMI_DEVICE_PARAMETER_BASEURL`
- [ ] Introduce a new method `QDMI_control_set_device_parameter` with semantics similar to `QDMI_control_set_job_parameter`

Feedback welcome!

**Comment by echavarria-mqv** [comment]
I may be missing something, but I think @denialhaag's last remark was not addressed with the proposed solutions:

> If a device needs a token only for job submission, it can, in principle, be set with QDMI_control_set_parameter. However, devices may also want to use a token for the query interface.

I find this particularly interesting because a _QDMI device_ could behave like a _QDMI client_ to interface with a real backend. Take, for example, Eviden's _QDMI device_. It should be capable of querying the noise model of a real backend, but we should not allow such an appliance to use all of the backend's _QDMI device_ query routines (let alone control routines).

**Comment by burgholzer** [comment]
> I may be missing something, but I think @denialhaag's last remark was not addressed with the proposed solutions:
> 
> > If a device needs a token only for job submission, it can, in principle, be set with QDMI_control_set_parameter. However, devices may also want to use a token for the query interface.

Technically, my proposal above was intended to address the full use case including the comment you quoted. However, as part of the discussion with @ystade on implementing that, it turned out that there's a bit of a bigger issue that we need to address.
This is probably also the thing you might be missing here: https://github.com/Munich-Quantum-Software-Stack/QDMI/issues/117, which is linked right above your comment.
There is some discussion to be had over there, but I think once resolved, it should also resolve this issue here.

> I find this particularly interesting because a _QDMI device_ could behave like a _QDMI client_ to interface with a real backend. Take, for example, Eviden's _QDMI device_. It should be capable of querying the noise model of a real backend, but we should not allow such an appliance to use all of the backend's _QDMI device_ query routines (let alone control routines).

Note that exposing access to the query methods isn't critical as they are read-only anyway, hence the name "query".

**Comment by echavarria-mqv** [comment]
I'm not sure I totally agree with this statement:

> Note that exposing access to the query methods isn't critical as they are read-only anyway, hence the name "query".

Even if we cannot think of specific scenarios now, some vendors might prefer exposing certain features/capabilities only to particular users.

**Comment by burgholzer** [comment]
> Even if we cannot think of specific scenarios now, some vendors might prefer exposing certain features/capabilities only to particular users.

Right 👍🏼. 
The solution proposed above facilitates that without adding anything extra though.
The Eviden appliance will need its own tokens for the devices it wants access to. 
These tokens might only offer limited access to the QDMI interface of the respective devices.
So I think once implemented, we are well covered even for this future scenario.

**Comment by echavarria-mqv** [comment]
> These tokens might only offer limited access to the QDMI interface of the respective devices.

Do you mean that the QDMI specification should assume that a _QDMI device_ has to control which routines are exposed to a _QDMI client_ based on the tokens it used for authentication?

**Comment by burgholzer** [comment]
> > These tokens might only offer limited access to the QDMI interface of the respective devices.
> 
> Do you mean that the QDMI specification should assume that a _QDMI device_ has to control which routines are exposed to a _QDMI client_ based on the tokens it used for authentication?

A device implementation does not _have to_, but it _might_ choose to do that.
At least this is one way that we can rather easily make this possible.

**Comment by ystade** [comment]
Just wanted to mention that there is already a `READONLY` and `READWRITE` mode to load a device provided from the driver side.

However, this is just a comment, and I totally agree with the already discussed approach to implement #117 properly and then this issue should also be resolved.
